### ISSUE TITLE
Warning about rolling back dirty (Server) object(s) after Spawner.start is called

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/jupyterhub#3830]
It seems that spawning a user in z2jh's CI system leads to a warning is emitted about state of possible relevance not being saved.

```
[D 2022-03-14 09:40:09.861 JupyterHub user:728] Calling Spawner.start for testuser-ba06d860-d846-4e6f-be49-2575356d9d3b
[W 2022-03-14 09:40:09.870 JupyterHub base:188] Rolling back dirty objects IdentitySet([<Server(0.0.0.0:8888)>])
```
_Link to github workflow logs [here](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/runs/5542660863?check_suite_focus=true#step:20:317)._

This is seen in logs of current z2jh main branch where KubeSpawner 3 hasn't yet been merged, even though it is also seen with KubeSpawner 3.

- jupyterhub 2.1.1 includes the warning: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/runs/4944776007?check_suite_focus=true#step:20:308
- jupyterhub 2.0.2 includes the warning: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/runs/4814486909?check_suite_focus=true#step:20:310
- jupyterhub 2.0.1 includes the warning: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/runs/4642738434?check_suite_focus=true#step:20:305
- jupyterhub 2.0.0 includes the warning: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/runs/5544971417?check_suite_focus=true#step:20:341

Logs are not available going back further though so I can't spot when its no longer around.

The question is: is this something to worry about, or expected and perhaps something not to flag as a warning if it is really expected?

**Comment by manics** [comment]
There's a report here of the database not matching the server:port returned by the spawner, maybe related? https://discourse.jupyter.org/t/jupyterhub-checks-for-notebook-servers-at-wrong-port-at-startup/13050

**Comment by minrk** [comment]
This warning does not seem to occur in the latest tests. I suspect this has been fixed by #3859. Mutating an ORM object without calling `db.commit()` is definitely wrong and would lead to warnings like this.

In general, dirty db is a cause for concern, because it means state has been attempted to insert into the db, but was not. However, if it was, as it appears, #3859, it should not be an issue in particular, because the state that would have been rolled back there was only _redundant_ URL info set while JupyterHub does not believe the server to be running yet.

What happened:

1. jupyterhub requests start
2. Spawner sets ip, port on ORM object as soon as it is known, but fails to commit (this is what gets rolled back)
3. Spawner _finishes_ starting, returning URL info
4. JupyterHub commits URL info to db and begins checking for availability before exposing it to the proxy

So JupyterHub does not attempt to _use_ the server URL until step 4, which is correctly inserted into the db. The 'dirty rollback' event only affects the value of the server between steps 2 and 3, during which the value is not used.

**Comment by consideRatio** [comment]
Thanks for sharing insights about this @minrk, closing this as resolved!
